File: 234-Palindrome Linked List.py
# 暴力 & 将值存在数组
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        tmp = []
        while head:
            tmp.append(head.val)
            head = head.next
        len_ = len(tmp)
        for i in range(int(len_ / 2)):
            if tmp[i] != tmp[len_ - i - 1]: return False
        return True

File: Algorithms/C++/234-Palindrome_Linked_List.cpp
// Runtime: 24 ms, faster than 66.39% of C++ online submissions for Palindrome Linked List.
// Memory Usage: 14.4 MB, less than 6.90% of C++ online submissions for Palindrome Linked List.
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        return check(head, head);
    }
private:
    bool check(ListNode*& head, ListNode* p) {
        if (!p)
            return true;
        bool is_pal = check(head, p->next);
        if (head->val != p->val)
            return false;
        head = head->next;
        return is_pal;
    }
};
